### ISSUE TITLE
Modules and layouts and themes (oh, my!)

### ISSUE THREAD

**Issue by BWPanda** [backdrop-ops/backdropcms.org#46]
As [I mentioned previously](https://github.com/backdrop-ops/backdropcms.org/issues/29#issuecomment-68507340), I think now that layouts (regions) are separate from themes, there should be a greater distinction between design (themes), layouts and functionality (modules).

A lot of people porting themes from Drupal to Backdrop are going to miss having theme-specific regions and are likely to bundle a custom layout with their theme. This, I believe, is supported and recommended, however I'd like to challenge this thinking with a slightly different proposal...

Say we have a Drupal theme (or WordPress theme) that has a specific colour-scheme, layout, font, slider, etc. and we want to port this to Backdrop. I recommend creating a:

1. Theme (with the appropriate colour-scheme, font, etc.)
2. Layout (with the necessary regions, etc.)
3. Module (with the code for the slider, if a compatible one doesn't already exist)

All of these should be available to download separately (since you won't always want the slider with the layout, or the layout with the theme), and then to provide a Backdrop-equivalent of the Drupal/WordPress theme, a fourth download would package together the other three: theme, layout & module.

This fourth download would be kind of like a Drupal distribution, except without core, install profiles, etc. I'd like to tentatively name this type of download a 'package'.

To reiterate, if someone wanted to make a Backdrop theme with specific regions, they'd make a Theme, make a Layout, and provide both together as a Package.

**Comment by BWPanda** [comment]
Then, on the eventual BackdropCMS.org project page for the package, it would list and link to the module(s), theme(s) and/or layout(s) that it bundles together. Also, each individual module/layout/theme project page could list the packages that include it, sorted by popularity.

**Comment by quicksketch** [comment]
> A lot of people porting themes from Drupal to Backdrop are going to miss having theme-specific regions and are likely to bundle a custom layout with their theme. This, I believe, is currently supported and recommended

I've heard a report that you can actually bundle a layout inside of a theme, but that's not actually intentional functionality. I wouldn't want to encourage that practice, but if it turns out we accidentally provided this, we may end up having to support it for 1.x. :frowning: 

I think this sounds like a good suggestion overall, however it's going to take a lot of infrastructure to make reality. We have a couple of items that would need to be done first, including converting BackdropCMS.org to Backdrop (https://github.com/backdrop-ops/backdropcms.org/issues/38) and build out the project repository of just "normal" projects based on Git repos: (https://github.com/backdrop-ops/backdropcms.org/issues/8).

**Comment by BWPanda** [comment]
Certainly; this was just a suggestion to discuss and consider while those other things happen.

**Comment by cellear** [comment]
Since layouts are new, maybe we could bundle this functionality with the distribution plans for them?  Let layouts identify recommended (or perhaps also required) modules and themes? Could layouts refuse to enable if they didn't have a required module or theme?

Adding a 4th category of downloadables to accommodate bundling seems like a pretty high cost.  I love the idea of Packages -- something less than distributions, but more inclusive than Features and Apps -- but that seems like a different issue.

**Comment by biolithic** [comment]
Luke, I think the point is:
Layout is the layout
Theme is the colors and skin
Modules are the functionality
Try not to make them break because of dependency.

I could be wrong.

For example, let's say I loved the Omega theme in Drupal (which really is a
set of layouts basically).  Ported, it is not really a theme but more of a
layout system and baseline.  I think a theme in Backdrop might be more like
"iOS8".  What's the grid system of iOS8?  It doesn't matter -- it makes
your headers and buttons and so forth look like iOS8.  Then, you will use
iOS8 with any grid/layout system in Backdrop.  (Note -- this is a fictional
example that I think the documentation is hinting towards).

Another example might be that this shows the "layout" --
http://thewebahead.net/61?gridset=show
While the theme might be header = grey, h1 = red big font, link color =
red, font family = some custom font etc...

**Comment by jenlampton** [comment]
@biolithic yes, exactly.

@cellear the whole point is to decouple layouts and themes. No theme should require (or recommend) a layout, and no layout should require (or recommend) a theme. Separate concepts, completely. All should work with all! Better UX for everyone :) (fewer broken sites)

**Comment by jenlampton** [comment]
I do think it would be a good idea to figure out some way to support packages. I'd also like to see packages be able to include default configurations that could be imported to set up things like content types and views. Then bam, you can have your slider too!

I'm not sure how this should be handled on Backdropcms.org. Perhaps a new contrib project of `type = package` or something?

**Comment by mikemccaffrey** [comment]
> something less than distributions, but more inclusive than Features

Well, perhaps we could modify the existing install profiles so they don't only have to run at site install. Then you can just run them on an existing site to install a group of themes and modules, and to automatically generate a bunch of structure items. 

The only downside, is that there'd be no way to disable one of these profiles and automatically remove all the items that have been created. Well, unless we added an uninstall hook or something.

**Comment by jenlampton** [comment]
> unless we added an uninstall hook or something.

That could work!  I like where this is going :)

**Comment by klonos** [comment]
A new project type of *profile* or *package* (to supplement modules , themes and layouts) with uninstall hook sounds like a great idea to me. It would be a form of "recipe" to install everything one needs in one go.

I think I've mentioned it previously in a related issue, but if we make this possible, then people producing tutorials or howto's could say "Download and install profile (or package) x" instead of "Download and install modules x, y and z. Then create content type a and view b and configure this and that setting." ...you get the point.

:+1:

**Comment by mikemccaffrey** [comment]
> A new project type of profile or package (to supplement modules, themes and layouts) with uninstall hook sounds like a great idea to me. It would be a form of "recipe" to install everything one needs in one go.

Well, install profiles pretty much already do that, so we probably don't need anything "new". We just need to provide the ability to add them to the contrib listings on the site.

**Comment by jenlampton** [comment]
I don't know if we should start with profiles, since those are installed when your site is installed and there's no going back afterwards.  I prefer the features analogy better - it's a suite of modules, themes, layouts and config, that can be enabled and disabled. There's a much larger use-case for features than there is for profiles, and I think that's what people will really want.

**Comment by mikemccaffrey** [comment]
Created backdrop/backdrop-issues/issues/1591 to discuss the relative merits of Features vs Profiles that can be run after site creation.

Let's limit this ticket to providing support on backdropcms.org for publishing contrib profiles, which seems like it needs to be done to support backdrop distrobutions, no matter what other additions/changes we make to how people can distribute packages of features.

**Comment by jenlampton** [comment]
I think we have a separate issue for that... let me see if i can find a link.

**Comment by BWPanda** [comment]
I've created a new issue for adding a new project type for 'recipes': https://github.com/backdrop/backdrop-issues/issues/3763

**Comment by jenlampton** [comment]
I don't think we should add these things to b.o until they are officially supported in core. For now, maybe we set up a new 'Recipies' section of the forum, and allow .zip uploads?
